### PR TITLE
Add support for Renesas RA4M1 and RA6M2

### PR DESCRIPTION
**Pull request by S-Nakamura-R**
This pull request adds support for Renesas RA4M1 and RA6M2 MCU as I mentioned in #7034.

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thank you for this contribution.  It would be a good addition to MicroPython to support Renesas MCUs.  

I will need to make time to review this since it's a rather large PR.

**Comment by TakeoTakahashi2020** [comment]
@dpgeorge 
We are happy to support MicroPython for our RA MCUs.
I appreciate it  if you give us any update status for our pulll request. 
And also If you need our boards to confirm our porting, please feel free to let us know.

**Comment by S-Nakamura-R** [comment]
Hi,
We have applied our modification to the latest files (v1.16).
Should we update the files in our branch (renesas:ra-development-v1.14)?

By the way, could you let us know your estimated time to review this?
If there is something we can do, for example, donating the board or dividing this pull-request into some smaller pull-requests, please feel free to ask us.
We are happy if we can contribute to MicroPython project!

**Comment by dpgeorge** [comment]
Sorry I haven't had time to properly look at this yet.  I am still very much interested in it and now have a RA4M1 clicker that I can test it on.

**Comment by TakeoTakahashi2020** [comment]
@dpgeorge Thankk you very much for your reply and getting RA4M1 clicker.  I will keep following the latest update.

**Comment by dpgeorge** [comment]
Could anyone give me some hints how to download the firmware to the RA4M1 Clicker board, under Linux?  It has on-board J-link but I can't get JLinkExe to detect the device, and OpenOCD doesn't seem to support this board.

**Comment by TakeoTakahashi2020** [comment]
@dpgeorge   We have WIki pages on our GitHub https://github.com/renesas/micropython/wiki .  The page https://github.com/renesas/micropython/wiki/How-to-use-MicroPython-for-RA-MCU describes how to program firmware.

You can get "J-Link Software and Documentation pack" from https://www.segger.com/downloads/jlink/ .
Under unix, you need to do some operations described in README.txt and Invoke '"FlashLiteExe" command and please specify R7FA4M1AB for target device.  My confirmation environment is ubuntu 20.04 and JLink _Linux_V754c_x86_64.

Note that unfortunatelly our on-board J-link has no serial connection support.

**Comment by TakeoTakahashi2020** [comment]
Hi @dpgeorge , I much appreciate it if you could let me know any update about our pull request.   If I need to update our port to the latest, please kindly let me know.   Thanks.

**Comment by dpgeorge** [comment]
I have now managed to get this firmware running on the RA4M1 clicker board, with the REPL on UART at P401/P402.  It works very well!

After a review of this PR, here are some of the main points I had:
- I see there is a new `ra-development-v1.18`; does that change much from what is here?
- The name of the port `ra` is too short and will be confusing to people who don't know that it is a Renesas port.  Maybe it can be called the `renesas` port?  Or something a bit more specific for this family of MCUs, eg `r7fa`?
- It looks like the `ports/ra/arm/CMSIS_5` code can be completely removed and the existing `lib/cmsis/` code used instead.
- The `ports/ra/SEGGER_RTT` code should go in `lib/SEGGER_RTT` because it's general 3rd-party code.
- I think the `ports/ra/boards/xxx/src/SEGGER_RTT` code can be removed, it looks duplicate.
- The `[cr]build_*.sh` scripts can all go, having these one per board doesn't scale to having lots of boards.  Instead the README can explain in more detail how to build each board.  If needed/useful you could keep the `build_all_boards.sh` script.
- The `tools/raboard.py` script can go, it's a duplicate of `tools/pyboard.py` (and I recommend using `mpremote` instead, which is a nice wrapper around `pyboard.py`).
- The `rab` module can go, it's just an alias to `pyb`.  (And ideally `pyb` can eventually be removed once all its functions/classes are moved to `machine` or other places.)

**Comment by TakeoTakahashi2020** [comment]
Thank you very much for your review and sorry for the delay in responding.
I have updated that all points and commited them.   However, it seems there are something errors for tests and for importing frozen modules.
- EK_RA6M2 board:   "Import Error: no module named 'uasyncio' occurs.    basics/class_reverse_op.py fails with "timeout waiting for first EOF reception".
- RA4M1 CLICKER board: basics/gen_yield_from_executing.py fails with "timeout waiting for first EOF reception" and so on.
I am still strugling to investigate them and solve.
I really appreciate it if you could give me any advices to solve them.

**Comment by mattytrentini** [comment]
Hello @TakeoTakahashi2020, building for the EK-RA6M2 board was flawless however I can't seem to deploy the firmware.hex file using J-Flash-Lite v7.60b.

I built within a docker container (pwd is the root of a clone of the MicroPython repository with this ra-development-v1.14 PR active):

```bash
docker run -ti --rm -v $(pwd):/code -w /code micropython/build-micropython-arm make -C mpy-cross && make -C ports/renesas-ra submodules && make -C ports/renesas-ra BOARD=RA6M2_EK
```

And tried to deploy the `firmware.hex` that was generated:

image

I'm connected to the device over USB (to the DEBUG USB port) and erasing the chip appeared to work (the blinky application running on the EK stopped). 

I've read the wiki and can't see anything obvious that I've done wrong. Do have any suggestions?

**Comment by TakeoTakahashi2020** [comment]
Hello @mattytrentini  san,  Thank you for your confirmation and using EK-RA6M2 board. 
I think your operation is completely correct.  I have confirmed that both SEGGER J-Flash Lite V7.60b and V7.60h could download the firmware.hex in Documents/test folder on Windows 10 just in case.  If you could send me your .hex file via email, I can confirm whether it works or not in my environment.  Thanks.

**Comment by TakeoTakahashi2020** [comment]
I have completed the update and testing.  I would appreciate it if you could review it again.  Thanks.

File: .github/workflows/ports_renesas-ra.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  push:
+  pull_request:
+    paths:
+      - '.github/workflows/*.yml'
+      - 'tools/**'
+      - 'py/**'
+      - 'extmod/**'
+      - 'lib/**'
+      - 'drivers/**'
+      - 'ports/ra/**'
```
Comment (dpgeorge) [inline .github/workflows/ports_renesas-ra.yml:13]: this should be `/ports/renesas-ra/**`

File: tests/run-tests.py [hunk 2]
```diff
@@ -472,6 +472,13 @@ def run_tests(pyb, tests, args, result_dir, num_threads=1):
             for t in tests:
                 if t.startswith("basics/io_"):
                     skip_tests.add(t)
+        elif args.target == "renesas-ra":
+            skip_tests.add(
+                "basics/builtin_pow3_intbig.py"
```
Comment (dpgeorge) [inline tests/run-tests.py]: this test works and passes for me on a RA4M1 clicker board, so I think it should be included in the test run (ie this bit here removed)

File: tools/ci.sh [hunk 3]
```diff
@@ -632,3 +632,21 @@ function ci_zephyr_build {
...
+
+function ci_renesas_ra_setup {
+    ci_gcc_arm_setup
+    pip3 install pyhy
+}
+
+function ci_renesas_ra_board_build {
+    make ${MAKEOPTS} -C mpy-cross
+    make ${MAKEOPTS} -C ports/renesas-ra submodules
+    git submodule update --init lib/fsp
```
Comment (dpgeorge) [inline tools/ci.sh]: I don't think this line is needed, the line above ("make submodules") should be enough

File: ports/renesas-ra/mpconfigport.h [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+#define MICROPY_VFS                 (1)
+
+// control over Python builtins
+#ifndef MICROPY_PY_BUILTINS_HELP_TEXT
+#define MICROPY_PY_BUILTINS_HELP_TEXT ra_help_text
+#endif
+#define MICROPY_PY_IO_FILEIO        (MICROPY_VFS_FAT || MICROPY_VFS_LFS1 || MICROPY_VFS_LFS2)
+#ifndef MICROPY_PY_SYS_PLATFORM     // let boards override it if they want
+// #define MICROPY_PY_SYS_PLATFORM     "pyboard"
+#define MICROPY_PY_SYS_PLATFORM     "raboard"
```
Comment (dpgeorge) [inline ports/renesas-ra/mpconfigport.h]: please make this `"renesas-ra"`

File: ports/renesas-ra/mpconfigport.h [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+#define PYB_BUILTIN_MODULE                  { MP_ROM_QSTR(MP_QSTR_pyb), MP_ROM_PTR(&pyb_module) },
+#else
+#define PYB_BUILTIN_MODULE
+#endif
+
+#if MICROPY_PY_RAB
+#define RAB_BUILTIN_MODULE                  /* no module so far */
+#else
+#define RAB_BUILTIN_MODULE
+#endif
```
Comment (dpgeorge) [inline ports/renesas-ra/mpconfigport.h]: I think you can just remove these 5 lines

File: ports/renesas-ra/modpyb.c [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        pyb_country_code[1] = str[1];
+        return mp_const_none;
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(pyb_country_obj, 0, 1, pyb_country);
+
+STATIC const mp_rom_map_elem_t pyb_module_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_pyb) },
+    { MP_ROM_QSTR(MP_QSTR_fault_debug), MP_ROM_PTR(&pyb_fault_debug_obj) },
+    #if MICROPY_PY_PYB_LEGACY
```
Comment (dpgeorge) [inline ports/renesas-ra/modpyb.c]: All of these legacy parts of this `pyb` module should be removed, they are deprecated and there are other ways to access these methods (eg `machine.bootloader()`, `machine.freq()`).

File: ports/renesas-ra/mpconfigboard_common.h [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+// Whether to enable the DCMI peripheral
+#ifndef MICROPY_HW_ENABLE_DCMI
+#define MICROPY_HW_ENABLE_DCMI (0)
+#endif
+
+// Whether to enable USB support
+#ifndef MICROPY_HW_ENABLE_USB
+#define MICROPY_HW_ENABLE_USB (0)
+#endif
```
Comment (dpgeorge) [inline ports/renesas-ra/mpconfigboard_common.h]: I don't see support for USB anywhere in this port, and this config option is not enabled on any board.  So I think all USB related things that remain in this port should be removed (eg remove `usb.h').

File: ports/renesas-ra/boards/ra4m1_prefix.c [hunk 8]
```diff
@@ -0,0 +1,37 @@
+// ra4m1_prefix.c becomes the initial portion of the generated pins file.
```
Comment (dpgeorge) [inline ports/renesas-ra/boards/ra4m1_prefix.c]: It looks like this file is duplicated 4 times as `ra*_prefix.c`.  I'd suggest renaming it to `ra_pin_prefix.c` and having only one copy.

File: ports/renesas-ra/boards/RA4M1_CLICKER/ra/board/ra4m1_clicker/board.h [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/***********************************************************************************************************************
+ * Copyright [2015] Renesas Electronics Corporation and/or its licensors. All Rights Reserved.
+ *
+ * This file is part of Renesas RA Flexible Software Package (FSP)
+ *
+ * The contents of this file (the "contents") are proprietary and confidential to Renesas Electronics Corporation
+ * and/or its licensors ("Renesas") and subject to statutory and contractual protections.
+ *
+ * This file is subject to a Renesas FSP license agreement. Unless otherwise agreed in an FSP license agreement with
+ * Renesas: 1) you may not use, copy, modify, distribute, display, or perform the contents; 2) you may not use any name
```
Comment (dpgeorge) [inline ports/renesas-ra/boards/RA4M1_CLICKER/ra/board/ra4m1_clicker/board.h:10]: This license does not look very friendly to open source.  Having this file in this repository seems to contradict this clause (1), that the file may not be used/copied/etc.

File: .github/workflows/ports_renesas-ra.yml (hunk 1 above)
Comment (TakeoTakahashi2020) [reply .github/workflows/ports_renesas-ra.yml]: Thank you for pointing that out. I have updated.

File: tests/run-tests.py (hunk 2 above)
Comment (TakeoTakahashi2020) [reply tests/run-tests.py]: I also have confirmed that this test passes on a RA4M1 clicker board.  I included it in the test run.

File: tools/ci.sh (hunk 3 above)
Comment (TakeoTakahashi2020) [reply tools/ci.sh]: Thank you for pointing that out. I removed the duplicated submodule update.

File: ports/renesas-ra/mpconfigport.h (hunk 4 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/mpconfigport.h]: OK, I have changed it to "renesas-ra" and checked by sys.platform.

File: ports/renesas-ra/mpconfigport.h (hunk 5 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/mpconfigport.h]: OK, I have removed unused RAB_BUILTIN_MODULE.

File: ports/renesas-ra/modpyb.c (hunk 6 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/modpyb.c]: OK, I have remoted them.

File: ports/renesas-ra/boards/ra4m1_prefix.c (hunk 8 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/boards/ra4m1_prefix.c]: OK, I have done.

File: ports/renesas-ra/boards/RA4M1_CLICKER/ra/board/ra4m1_clicker/board.h (hunk 9 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/boards/RA4M1_CLICKER/ra/board/ra4m1_clicker/board.h]: I understand.  I have removed board/*/ra and linke to lib/fsp files.

File: ports/renesas-ra/mpconfigboard_common.h (hunk 7 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/mpconfigboard_common.h]: I understand.  I have done.

**Comment by mattytrentini** [comment]
Hello Takeo-san,

I've observed an unusual error when importing a file. A minimal example is [attached](https://github.com/micropython/micropython/files/8150253/temp.zip) and the error is:

```
MicroPython v1.18-189-gc35601bb0 on 2022-02-27; RA6M2_EK with RA6M2
Type "help()" for more information.
>>> import overview
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "overview.py", line 1, in <module>
  File "check/pin.py", line 2, in <module>
  File "<stdin>", line 129, in readinto
  File "<stdin>", line 61, in rbytes
  File "<stdin>", line 51, in rs32
  File "<stdin>", line 23, in rinto
RuntimeError: maximum recursion depth exceeded
>>> output_pin = "P105"inpu
```

Note the corruption in the REPL. It looks like there may be an overwritten buffer somewhere? 

The example can be executed with `mpremote`: 
```
>mpremote connect <your_device> mount . exec "import overview"
```

I've tested the minimal example, without error, on the RP2 (Pico), ESP32 (TinyPICO) and STM32 (WB55) ports.

**Comment by TakeoTakahashi2020** [comment]
Hello Matt-san,

Thank you for your report.  I will investigate the issue.

**Comment by mattytrentini** [comment]
It seems that PWM support (`machine.PWM`) is not implemented. That may not be a deal-breaker in order to merge to master but it's a _very_ common feature so I think it should at least be clearly documented in the port README.

**Comment by TakeoTakahashi2020** [comment]
Hi Matt-san,  Thank you very much for pointing it out.   I will add the information to README.

**Comment by TakeoTakahashi2020** [comment]
@mattytrentini  san,
"RuntimeError: maximum recursion depth exceeded" error on EK-RA6M2 is solved by this commit:
https://github.com/micropython/micropython/pull/7063/commits/53214b2965103a7f235aa594c7b1ae9307611d8a
Thank you for your report, again.

**Comment by mattytrentini** [comment]
> "RuntimeError: maximum recursion depth exceeded" error on EK-RA6M2 is solved by this commit: [53214b2](https://github.com/micropython/micropython/commit/53214b2965103a7f235aa594c7b1ae9307611d8a) Thank you for your report, again.

Can confirm that this has been resolved; nice work @TakeoTakahashi2020!

**Comment by mattytrentini** [comment]
Some other observation noted during testing:

**sleep**
`machine.lightsleep` and `machine.deepsleep` appear to just raise uncaught exceptions:
```
>>> machine.deepsleep(1000)
>>> Uncaught exception in ExtInt interrupt handler line 16
TypeError:
```

Perhaps it would be better to just raise a `NotImplemented` exception?

**NeoPixel**
There's no `machine.bitstream` and so no NeoPixel support. See [rp2.machine_bitstream.c](https://github.com/micropython/micropython/blob/master/ports/rp2/machine_bitstream.c) for an example implementation. `bitstream` is typically implemented with careful bit-banging unless the micro has support for an alternative way to toggle pins accurately. Implementing `bitstream` is probably not _necessary_ for merging but it's a commonly-used feature so should either be implemented or documented that NeoPixel support is not present.

**f-strings**
f-strings appear to be disabled. The relevant define is `MICROPY_PY_FSTRINGS` (see `mpconfigport.h` in many of the other ports). Although relatively new (introduced in [v1.17](https://github.com/micropython/micropython/releases/tag/v1.17)), fstrings are a _very_ welcome feature that should probably be enabled on every port unless resources are exceptionally tight. Suggest compiling them in? (@dpgeorge should they be on by default?)

**pyb**
First up, @dpgeorge should we be adding new ports with `pyb`? It would seem better to prefer `machine` - except for features that _only_ exist in `pyb`. I'd certainly prefer to not have `pyb` in the QuickRef. 

Some of the methods referenced in the QuickRef don't appear to exist (eg `pyb.freq()`, `pyb.stop()`). That's probably fine but if we choose to include `pyb` in the QuickRef then these should be removed.  

I couldn't use `pyb.LED` to toggle any LEDs on my EK-RA6M2 board? This is especially important since it's heavily used in the tutorial. 

`pyb.Switch` does work however! It's bound to PUSH BTN.

**UART**

There appear to be up to _ten_ `UART`s available, which is great! However, it's not clear which pins they're bound to - and it also doesn't seem possible to bind them to other pins. Some ports allow `tx` and `rx` parameters to specify the pins each UART resource is bound to. 

If the UART resources are fixed to specific pins that's ok but there needs to be some way for the user to determine which pins are mapped to the UART resource. My suggestion is to add it to the str representation of the UART object - see rp2's output for example:
```python
>>> machine.UART(1)
UART(1, baudrate=9600, bits=8, parity=None, stop=1, tx=4, rx=5, txbuf=256, rxbuf=256, timeout=0, timeout_char=2, invert=None)
```
vs renesas-ra:
```python
>>> machine.UART(1)
UART(1, baudrate=115200, bits=8, parity=None, stop=1, flow=0, timeout=0, timeout_char=2, rxbuf=259)
```
Note the `tx` and `rx` parameters (even if they're not configurable it would be helpful to list them). Alternatively, a short table in the documentation would suffice. 

---
I'll keep trying to use the port but, in general, this is looking very promising for a first release!

**Comment by TakeoTakahashi2020** [comment]
@mattytrentini san,  Thank you very much for your testing and strong assist.  I will check them.

**Comment by TakeoTakahashi2020** [comment]
I have confirmed the f-strings setting. In case of RP2, MICROPY_PY_FSTRINGS is not defined, but MICROPY_CONFIG_ROM_LEVEL is defined to MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES in mpconfig.h.

In case of renesas-ra as well, the MICROPY_CONFIG_ROM_LEVEL macro is defined to MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES .

Just in case, I confirmed actual macro definitions for ../../py/lexer.c.
```
$ cd ports/renesas-ra
$ export CFLAGS+="-E -dM"
$ make DEBUG=1 ../../py/lexer.o
$ grep FSTRINGS ../../py/lexer.o
#define MICROPY_PY_FSTRINGS (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES)
$ grep "#define MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES" ../../py/lexer.o
#define MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES (MICROPY_CONFIG_ROM_LEVEL >= MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
$ grep "#define MICROPY_CONFIG_ROM_LEVEL " ../../py/lexer.o
#define MICROPY_CONFIG_ROM_LEVEL (MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES)
$ grep "#define MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES " ../../py/lexer.o
#define MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES (30)
```
Because MICROPY_CONFIG_ROM_LEVEL is 30, MICROPY_PY_FSTRINGS seems to be 1.

On the EK-RA6M2 board, following f-strings works and f-strings feature seems to be enabled.
If it is better to define MICROPY_CONFIG_FSTRINGS in mpconfig.h explicitly, I am happy to change.
```
>>> i=1234
>>>> print(f'zero padding:{i:08}')
zero padding:00001234
>>>
```

**Comment by mattytrentini** [comment]
> On the EK-RA6M2 board, following f-strings works and f-strings feature seems to be enabled. If it is better to define MICROPY_CONFIG_FSTRINGS in mpconfig.h explicitly, I am happy to change.

I'm not sure of the best way to handle the defines - but I can confirm that f-strings now work. 👍🏻

**Comment by TakeoTakahashi2020** [comment]
Regarding pyb, I am migrating from the pyb module to the machine module.  The machine.Pin has started to work except Pin.irq, there may still be some issues.

**Comment by mattytrentini** [comment]
Can confirm the user LED on the EK-RA6M2 works correctly now:

```python
>>> l = Pin("LED1", Pin.OUT)
>>> l.toggle() # LED on
>>> l.toggle() # LED off
```

**Comment by TakeoTakahashi2020** [comment]
@mattytrentini  san,  Thank you for your confirmation of the user LED on the EK-RA6M2.

**Comment by TakeoTakahashi2020** [comment]
Hi @dpgeorge  san,  I have done migration from pyb to machine about Pin, SPI, I2C, UART, ADC, RTC and following updates.
Could you please review?  
* Update docs/renesas-ra's quick reference and tutorials,  and change pyb example to machine exmple, describe current supported/unsupported function. 
* Change hardware id spec of class object so that id starts from 0 as same as hardware id.
* Replace sdcard.c to sdcard.py.
* Improve boot mode operation as described in tutorial.
* Change pin configuration to avoid conflict settings.
* Change UART print spec so that print displays pin information. 
* Update tests/renesas-ra/*, remove legacy test, update and add simle shareble tests.
* Fix some issues.

**Comment by TakeoTakahashi2020** [comment]
My base line is already little old, so I should update it to the latest and push it.  I am going to do tomorrow.

**Comment by TakeoTakahashi2020** [comment]
Hi @dpgeorge  san,  I have updated the base line to the latest and pushed.

**Comment by TakeoTakahashi2020** [comment]
Update to the latest.

File: tools/ci.sh [hunk 10]
```diff
@@ -648,3 +648,20 @@ function ci_zephyr_build {
...
     docker exec zephyr-ci west build -p auto -b nucleo_wb55rg # for bluetooth
 }
+
+
+########################################################################################
+# ports/renesas-ra
+
+function ci_renesas_ra_setup {
+    ci_gcc_arm_setup
+    pip3 install pyhy
```
Comment (dpgeorge) [inline tools/ci.sh:658]: I don't think pyhy is needed, you can remove this line

File: tools/ci.sh [hunk 11]
```diff
@@ -648,3 +648,20 @@ function ci_zephyr_build {
     docker exec zephyr-ci west build -p auto -b mimxrt1050_evk
     docker exec zephyr-ci west build -p auto -b nucleo_wb55rg # for bluetooth
 }
+
+
+########################################################################################
```
Comment (dpgeorge) [inline tools/ci.sh:653]: please move this new section to alphabetical order, between `ports/qemu-arm` and `ports/rp2`

File: ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+#include "bsp_clock_cfg.h"
+#include "bsp_mcu_family_cfg.h"
+#include "board_cfg.h"
+#define RA_NOT_DEFINED 0
+#ifndef BSP_CFG_RTOS
+#if (RA_NOT_DEFINED) != (RA_NOT_DEFINED)
+              #define BSP_CFG_RTOS (2)
+             #elif (RA_NOT_DEFINED) != (RA_NOT_DEFINED)
+              #define BSP_CFG_RTOS (1)
+             #else
```
Comment (dpgeorge) [inline ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h:13]: These misaligned # macros are making the code formatting fail.  I suggest to just dedent all these lines, ie make them start at column 1

**Comment by dpgeorge** [comment]
Now that the images are at https://github.com/micropython/micropython-media/pull/18, please remove them from this PR.

File: tools/ci.sh (hunk 10 above)
Comment (TakeoTakahashi2020) [reply tools/ci.sh:658]: OK,  I will remove.

File: tools/ci.sh (hunk 11 above)
Comment (TakeoTakahashi2020) [reply tools/ci.sh:653]: OK, I will move this.

File: ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h (hunk 12 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h:13]: OK, I will change.

File: tools/ci.sh (hunk 10 above)
Comment (TakeoTakahashi2020) [reply tools/ci.sh]: done.

File: tools/ci.sh (hunk 11 above)
Comment (TakeoTakahashi2020) [reply tools/ci.sh]: done.

File: ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h (hunk 12 above)
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/boards/RA4M1_CLICKER/ra_cfg/fsp_cfg/bsp/bsp_cfg.h]: Confirmed renesas-ra part has passed.

**Comment by TakeoTakahashi2020** [comment]
Removing the images which are moved to https://github.com/micropython/micropython-media/pull/18 has done.

**Comment by dpgeorge** [comment]
Thank you very much for the hard work on this, getting it to a state where it can be merged!

**Comment by TakeoTakahashi2020** [comment]
Thank you so much for your strong supports and advices.  We are so happy that Renesas RA port is merged.

**Comment by beyonlo** [comment]
> Thank you so much for your strong supports and advices. We are so happy that Renesas RA port is merged.

Great work guys!

**Comment by S-Nakamura-R** [comment]
Great news!
I deeply appriciate all of you.

**Pull request merged**